gossip/src/crds_value.rs

```diff
@@ -490,84 +496,193 @@ impl Sanitize for NodeInstance {
     }
 }
 
-#[derive(Serialize, Deserialize, Clone, Default, PartialEq, Eq, AbiExample, Debug)]
+#[derive(Serialize, Deserialize, Clone, Debug, PartialEq, Eq, AbiExample, AbiEnumVisitor)]
+pub enum SlotsOffsets {
+    RunLengthEncoding(RunLengthEncoding),
+    RawOffsets(RawOffsets),
+}
+
+#[derive(Deserialize, Serialize, Clone, Debug, PartialEq, Eq, AbiExample)]
+struct U16(#[serde(with = "serde_varint")] u16);
+
+#[derive(Deserialize, Serialize, Clone, Debug, PartialEq, Eq, AbiExample)]
+pub struct RunLengthEncoding {
+    // The vector always starts with 1. Encode number of 1's and 0's consecutively.
+    // For example, 110000111 is [2, 4, 3].
+    encoded: Vec<U16>,
+}
+
+impl RunLengthEncoding {
+    // Return whether the encoded vec is already full.
+    fn valid_and_has_space(total_serde_bytes: &mut usize, count: usize) -> bool {
+        match u16::try_from(count) {
+            Ok(current_bit_count) => {
+                let serde_bits = u16::BITS - current_bit_count.leading_zeros();
+                let serde_bytes = ((serde_bits + 6) / 7).max(1) as usize;
+                if *total_serde_bytes + serde_bytes > RestartLastVotedForkSlots::MAX_SPACE {
+                    false
+                } else {
+                    *total_serde_bytes += serde_bytes;
+                    true
+                }
+            }
+            Err(e) => {
+                info!("Failed to convert count to u16: {e:?}");
+                false
+            }
+        }
+    }
+
+    pub fn new(bits: &BitVec<u8>) -> Self {
+        let mut total_serde_bytes = 0;
+        let encoded = (0..bits.len())
+            .map(|i| bits.get(i))
+            .dedup_with_count()
+            .map(|(count, _)| count)
+            .take_while(|count| Self::valid_and_has_space(&mut total_serde_bytes, *count))
+            .map(|count| U16(count.try_into().unwrap()))
+            .collect();
+        Self { encoded }
+    }
+
+    pub fn slots_count(&self) -> usize {
+        self.encoded.iter().map(|x| x.0 as usize).sum::<usize>()
+    }
+
+    pub fn to_slots(&self, last_slot: Slot, min_slot: Slot) -> Vec<Slot> {
+        let mut result: Vec<Slot> = self
+            .encoded
+            .iter()
+            .zip([1, 0].iter().cycle())
+            .flat_map(|(bit_count, bit)| std::iter::repeat(bit).take(bit_count.0.into()))
```

Comment:
`.into` is ambiguous that what type change happens here.
can you use `usize::from` here?

Reply:
Done.